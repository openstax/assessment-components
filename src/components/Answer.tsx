import cn from 'classnames';
import React, { ReactNode } from 'react';
import { ALPHABET, isAnswerChecked, isAnswerCorrect, isAnswerIncorrect } from '../utils';
import { answerType, chosenAnswerType, idType } from '../types';

export interface AnswerProps {
  answer: answerType;
  iter: number;
  qid: idType;
  type: 'teacher-review' | 'teacher-preview' | 'student' | 'student-mpp';
  hasCorrectAnswer: boolean;
  onChangeAnswer: () => void;
  disabled: boolean;
  chosenAnswer: chosenAnswerType;
  correctAnswerId?: idType;
  incorrectAnswerId?: idType;
  onKeyPress: () => void;
  answered_count?: number;
  correctIncorrectIcon?: ReactNode,
  feedback?: ReactNode;
  radioBox?: ReactNode;
  contentRenderer?: JSX.Element;
}

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
type ComponentType = keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>;

interface ContentRendererProps<T extends ComponentType | undefined> {
  className?: string;
  component?: T extends undefined ? undefined :
    T extends ComponentType ? React.ReactComponentElement<T>:
    never;
  html: string;
}

const AnswerContent = (<T extends ComponentType | undefined>(
  {html, component, ...props}: ContentRendererProps<T>
) => {
  if (component !== undefined) {
    return React.cloneElement(component, {html, ...props});
  }
  return <div dangerouslySetInnerHTML={{ __html: html }} {...props} />;
});

export const Answer = (props: AnswerProps) => {
  const {
    type,
    iter,
    answer,
    disabled,
    onKeyPress,
    qid,
    chosenAnswer,
    correctAnswerId,
    incorrectAnswerId,
    hasCorrectAnswer,
    answered_count,
    feedback,
    contentRenderer
  } = props;

  let body, selectedCount;

  const isChecked = isAnswerChecked(answer, chosenAnswer);
  const isCorrect = isAnswerCorrect(answer, correctAnswerId);
  const isIncorrect = isAnswerIncorrect(answer, incorrectAnswerId);

  const correctIncorrectIcon = (
    <div className="correct-incorrect">
      {isCorrect && props.correctIncorrectIcon}
    </div>
  );

  const classes = cn('answers-answer', {
    'disabled': disabled,
    'answer-checked': isChecked,
    'answer-correct': isCorrect && type !== 'student-mpp',
    'answer-incorrect': incorrectAnswerId && isAnswerIncorrect(answer, incorrectAnswerId),
  });

  let ariaLabel = `${isChecked ? 'Selected ' : ''}Choice ${ALPHABET[iter]}`;
  // somewhat misleading - this means that there is a correct answer,
  // not necessarily that this answer is correct
  if (hasCorrectAnswer) {
    ariaLabel += `(${isCorrect ? 'Correct' : 'Incorrect'} Answer)`;
  }
  ariaLabel += ':';

  let onChangeAnswer, radioBox;

  if (!hasCorrectAnswer
    && (type !== 'teacher-review')
    && (type !== 'teacher-preview')
    && (type !== 'student-mpp')) {
    ({ onChangeAnswer } = props);
  }

  if (onChangeAnswer) {
    radioBox = (
      <input
        type="radio"
        className="answer-input-box"
        checked={isChecked}
        id={`${qid}-option-${iter}`}
        name={`${qid}-options`}
        onChange={onChangeAnswer}
        disabled={disabled}
      />
    );
  }

  if (type === 'teacher-review') {
    let percent = 0;
    if (answer.selected_count && answered_count) {
      percent = Math.round((answer.selected_count / answered_count) * 100);
    }
    selectedCount = (
      <span
        className="selected-count"
        data-percent={`${percent}`}
      >
        {answer.selected_count}
      </span>
    );

    body = (
      <div className="review-wrapper">
        <div className={cn('review-count', { 'green': isCorrect, 'red': !isCorrect })}>
          {selectedCount}
          <span className={cn('letter', { 'green': isCorrect, 'red': !isCorrect })}>
            {ALPHABET[iter]}
          </span>
        </div>

        <div className="answer-answer">
          <div className="answer-content">
            <AnswerContent className="answer-content" component={contentRenderer} html={answer.content_html} />
          </div>
          {feedback}
        </div>
      </div>
    );
  } else {
    body = (
      <>
        {type === 'teacher-preview' && correctIncorrectIcon}
        {selectedCount}
        {radioBox}
        <label
          onKeyPress={onKeyPress}
          htmlFor={`${qid}-option-${iter}`}
          className="answer-label">
          <span className="answer-letter-wrapper">
            <button
              onClick={onChangeAnswer}
              aria-label={ariaLabel}
              className="answer-letter"
              disabled={disabled || isIncorrect}
              data-test-id={`answer-choice-${ALPHABET[iter]}`}
            >
              {ALPHABET[iter]}
            </button>
          </span>
          <div className="answer-answer">
            <AnswerContent className="answer-content" component={contentRenderer} html={answer.content_html} />
            {feedback}
          </div>
        </label>
      </>
    );
  }

  return (
    <div className="openstax-answer">
      <section role="region" className={classes}>
        {body}
      </section>
    </div>
  );
}
Answer.displayName = 'OSAnswer';
