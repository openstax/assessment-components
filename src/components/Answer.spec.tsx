import { Answer } from './Answer';
import renderer from 'react-test-renderer';

describe('Answer', () => {
  const props = {
    iter: 1,
    answer: {
      id: 1,
      correctness: '',
      isCorrect: true,
      content_html: 'Some great content',
    },
    onChangeAnswer: () => jest.fn(),
    disabled: false,
    onKeyPress: () => jest.fn(),
    qid: 1,
    hasCorrectAnswer: false,
    chosenAnswer: [''] ,
    correctAnswerId: 2,
    incorrectAnswerId: 0,
    answered_count: 0,
    show_all_feedback: false,
    keyControl: '',
    selectedCount: 1,
    isIncorrect: false,
    isCorrect: false
  };

  it('matches snapshot', () => {
     const tree = renderer.create(
      <Answer type='student' {...props} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with a custom renderer', () => {
    props.answer.content_html = `<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
`;
    const tree = renderer.create(
      <Answer type='student' {...props}>
        <div dangerouslySetInnerHTML={{ __html: props.answer.content_html }}></div>
      </Answer>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
