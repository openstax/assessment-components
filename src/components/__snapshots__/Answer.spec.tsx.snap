// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Answer matches snapshot 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer"
  >
    <input
      checked={false}
      className="answer-input-box"
      disabled={false}
      id="1-option-1"
      name="1-options"
      onChange={[Function]}
      type="radio"
    />
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        aria-label="Choice B:"
        className="answer-letter-wrapper"
        data-answer-choice="B"
        data-test-id="answer-choice-B"
      />
      <div
        className="answer-answer"
      >
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </label>
  </section>
</div>
`;

exports[`Answer renders a checked answer 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer answer-selected"
  >
    <input
      checked={true}
      className="answer-input-box"
      disabled={false}
      id="1-option-1"
      name="1-options"
      onChange={[Function]}
      type="radio"
    />
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        aria-label="Selected Choice B:"
        className="answer-letter-wrapper"
        data-answer-choice="B"
        data-test-id="answer-choice-B"
      />
      <div
        className="answer-answer"
      >
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </label>
  </section>
</div>
`;

exports[`Answer renders a correct answer 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer disabled answer-correct"
  >
    <input
      checked={false}
      className="answer-input-box"
      disabled={true}
      id="1-option-1"
      name="1-options"
      onChange={[Function]}
      type="radio"
    />
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        aria-label="Choice B:"
        className="answer-letter-wrapper"
        data-answer-choice="B"
        data-test-id="answer-choice-B"
      />
      <div
        className="answer-answer"
      >
        <div
          className="sc-jSMfEi ecLwSd"
        >
          Correct
           Answer
        </div>
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </label>
  </section>
</div>
`;

exports[`Answer renders an incorrect answer 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer answer-incorrect"
  >
    <input
      checked={false}
      className="answer-input-box"
      disabled={true}
      id="1-option-1"
      name="1-options"
      onChange={[Function]}
      type="radio"
    />
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        aria-label="Choice B:"
        className="answer-letter-wrapper"
        data-answer-choice="B"
        data-test-id="answer-choice-B"
      />
      <div
        className="answer-answer"
      >
        <div
          className="sc-jSMfEi faCGeG"
        >
          Incorrect
           Answer
        </div>
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </label>
  </section>
</div>
`;

exports[`Answer renders feedback 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer"
  >
    <input
      checked={false}
      className="answer-input-box"
      disabled={false}
      id="1-option-1"
      name="1-options"
      onChange={[Function]}
      type="radio"
    />
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        aria-label="Choice B:"
        className="answer-letter-wrapper"
        data-answer-choice="B"
        data-test-id="answer-choice-B"
      />
      <div
        className="answer-answer"
      >
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
        <aside>
          <div
            className="question-feedback-content has-html"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<div>Insightful commentary</div>",
              }
            }
          />
        </aside>
      </div>
    </label>
  </section>
</div>
`;

exports[`Answer renders teacher preview 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer answer-correct"
  >
    <div
      className="correct-incorrect"
    >
      <span>
        Iconic
      </span>
    </div>
    <input
      checked={false}
      className="answer-input-box"
      disabled={true}
      id="1-option-1"
      name="1-options"
      onChange={[Function]}
      type="radio"
    />
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        aria-label="Choice B:"
        className="answer-letter-wrapper"
        data-answer-choice="B"
        data-test-id="answer-choice-B"
      />
      <div
        className="answer-answer"
      >
        <div
          className="sc-jSMfEi ecLwSd"
        >
          Correct
           Answer
        </div>
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </label>
  </section>
</div>
`;

exports[`Answer renders teacher review 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer"
  >
    <div
      className="review-wrapper"
    >
      <div
        className="review-count red"
      >
        <span
          className="selected-count"
          data-percent="50"
        >
          5
        </span>
        <span
          className="letter red"
        >
          B
        </span>
      </div>
      <div
        className="answer-answer"
      >
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </div>
  </section>
</div>
`;

exports[`Answer renders with a custom renderer if set 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer"
  >
    <input
      checked={false}
      className="answer-input-box"
      disabled={false}
      id="1-option-1"
      name="1-options"
      onChange={[Function]}
      type="radio"
    />
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        aria-label="Choice B:"
        className="answer-letter-wrapper"
        data-answer-choice="B"
        data-test-id="answer-choice-B"
      />
      <div
        className="answer-answer"
      >
        <div
          className="custom-renderer"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </label>
  </section>
</div>
`;
